File: fsharp/binary-search-tree/BinarySearchTree.fs
module BinarySearchTree

type Node = { value: int; left: Node option; right: Node option }

let rec insert (tree: Node option) (value: int): Node =
    match tree with
    | None -> { value=value; left=None; right=None; }
    | Some t when value > t.value -> { t with right=Some (insert t.right value) }
    | Some t -> { t with left=Some (insert t.left value) }

let rec sort (tree: Node option): int list =
    match tree with
    | None -> []
    | Some node -> (sort node.left) @ [node.value] @ (sort node.right)

let left node = node.left

let right node = node.right

let data node = node.value

let create (items: int list): Node = 
    items
    |> List.fold (fun acc item -> Some (insert acc item)) None
    |> Option.get

let sortedData node = node |> Some |> sort

File: kotlin/binary-search-tree/src/main/kotlin/BinarySearchTree.kt
import java.util.*

class BinarySearchTree<T : Comparable<T>> {

    data class Node<T>(val data: T, val left: Node<T>? = null, val right: Node<T>? = null)

    var root: Node<T>? = null

    fun insert(value: T) {
        root = insert(value, root)
    }

    private fun insert(value: T, node: Node<T>?): Node<T> = when {
        node == null -> Node(value)
        value > node.data -> node.copy(right = insert(value, node.right))
        else -> node.copy(left = insert(value, node.left))
    }

    fun asSortedList(): List<T> = asSortedList(root)

    private fun asSortedList(node: Node<T>?): List<T> {
        if (node == null) {
            return emptyList()
        }

        return asSortedList(node.left) + node.data + asSortedList(node.right)
    }

    fun asLevelOrderList(): List<T> {
        val items = mutableListOf<T>()
        val queue = LinkedList<Node<T>?>()
        queue.add(root)

        while (queue.isNotEmpty()) {
            val node = queue.poll() ?: continue

            items.add(node.data)
            queue.add(node.left)
            queue.add(node.right)
        }

        return items
    }
}
